dcpy/connectors/edm/packaging.py

```diff
@@ -0,0 +1,150 @@
+import os
+import shutil
+from pathlib import Path
+from typing import Callable
+from dataclasses import dataclass
+import typer
+
+from dcpy.utils import s3
+from dcpy.utils.logging import logger
+from dcpy.connectors.edm import publishing
+
+BUCKET = "edm-distributions"
+BUCKET_ACL = s3.string_as_acl("public-read")
+
+DOWNLOAD_ROOT_PATH = Path(".library")
+OUTPUT_ROOT_PATH = Path(".output")
+
+
+@dataclass
+class PackageKey(publishing.ProductKey):
+    product: str
+    version: str
+
+    def __str__(self):
+        return f"{self.product} - {self.version}"
+
+    @property
+    def path(self) -> str:
+        return f"{self.product}/{self.version}"
+
+
+def colp(publish_key: publishing.PublishKey, package_key: PackageKey):
+    raise NotImplementedError
+
+
+def templatedb(publish_key: publishing.PublishKey, package_key: PackageKey):
+    output_path = OUTPUT_ROOT_PATH / package_key.path
+    shutil.copytree(
+        DOWNLOAD_ROOT_PATH / publish_key.path,
+        output_path,
+        dirs_exist_ok=True,
+    )
+    shutil.copy(
+        output_path / "templatedb.csv",
+        output_path / "templatedb_packaged.csv",
+    )
+
+
+@dataclass
+class PackageMetadata:
+    product: str
+    packaged_name: str
+    packaging_function: Callable
+
+
+DATASET_PACKAGE_METADATA = {
+    "db-colp": PackageMetadata("db-colp", "dcp_colp", colp),
+    "db-template": PackageMetadata("db-colp", "dcp_template_db", templatedb),
+}
+
+
+def upload(package_key: PackageKey) -> None:
+    local_folder_path = OUTPUT_ROOT_PATH / package_key.path
+    meta = s3.generate_metadata()
```

Comment:
Would we maybe want to use existing metadata? In this case, we're losing the metadata of the s3 files in publishing (losing as in not shuttling them along)

Reply:
so using `s3.get_metadata` when data is downloaded and passing it in into this upload function as an arg?

Reply:
That's roughly what I was getting at. Definitely open for discussion though

Reply:
It might make more sense to use `build_metadata.json` as source of truth than s3 file metadata though

Reply:
Which we are generating for all products using recipe now

Reply:
doesn't look like the entirety of `build_metadata.json` can become s3 file metadata because it has dictionaries as values ([run here](https://github.com/NYCPlanning/data-engineering/actions/runs/7350128126/job/20011262884#step:6:1619)). I confirmed that the nested dicts cause this by running tests locally

also using `generate_metadata()` may be more desirable so that the packaged files reflect the datetime, commit, and action run that actually packaged them. plus `build_metadata.json` will still be a file with that packaged data

Reply:
That makes sense. I'm good with that.

With build_metadata.json, we'd def want to drop the recipe, that's what's causing the issue.

Reply:
But for now seems we don't necessarily need anything from it

Reply:
sounds good! I'll drop/revise the relevant commits c82fe4e and 057e692